state/protocol/badger/mutator.go

```diff
@@ -526,99 +527,55 @@ func (m *FollowerState) Finalize(ctx context.Context, blockID flow.Identifier) e
-	// Service notifications and updating metrics happen when we finalize the _first_
-	// block of the new Epoch (same convention as for Epoch-Phase-Changes)
-	// Approach: We retrieve the parent block's epoch information. If this block's view
-	// exceeds the final view of its parent's current epoch, this block begins the next epoch.
-	parentBlocksEpoch := m.AtBlockID(header.ParentID).Epochs().Current()
-	parentEpochFinalView, err := parentBlocksEpoch.FinalView()
-	if err != nil {
-		return fmt.Errorf("could not get parent epoch final view: %w", err)
-	}
-
-	// When this block's view exceeds the parent epoch's final view, this block
-	// represents the first block of the next epoch. Therefore we update metrics
-	// related to the epoch transition here.
-	//
-	// We skip updating these metrics when EECC has been triggered
-	if header.View > parentEpochFinalView && !epochFallbackTriggered {
-		events = append(events, func() { m.consumer.EpochTransition(currentEpochSetup.Counter, header) })
+	// Determine metric updates and protocol events related to epoch phase
+	// changes and epoch transitions.
+	// If epoch emergency fallback is triggered, the current epoch continues until
+	// the next spork - so skip these updates.
+	if !epochFallbackTriggered {
+		epochPhaseMetrics, epochPhaseEvents, err := m.epochPhaseMetricsAndEventsOnBlockFinalized(header, epochStatus)
+		if err != nil {
+			return fmt.Errorf("could not determine epoch phase metrics/events for finalized block: %w", err)
+		}
+		metrics = append(metrics, epochPhaseMetrics...)
+		events = append(events, epochPhaseEvents...)
 
-		// set current epoch counter corresponding to new epoch
-		events = append(events, func() { m.metrics.CurrentEpochCounter(currentEpochSetup.Counter) })
-		// set epoch phase - since we are starting a new epoch we begin in the staking phase
-		events = append(events, func() { m.metrics.CurrentEpochPhase(flow.EpochPhaseStaking) })
-		// set current epoch view values
-		events = append(
-			events,
-			func() { m.metrics.CurrentEpochFinalView(currentEpochSetup.FinalView) },
-			func() { m.metrics.CurrentDKGPhase1FinalView(currentEpochSetup.DKGPhase1FinalView) },
-			func() { m.metrics.CurrentDKGPhase2FinalView(currentEpochSetup.DKGPhase2FinalView) },
-			func() { m.metrics.CurrentDKGPhase3FinalView(currentEpochSetup.DKGPhase3FinalView) },
-		)
+		epochTransitionMetrics, epochTransitionEvents, err := m.epochTransitionMetricsAndEventsOnBlockFinalized(header, currentEpochSetup)
+		if err != nil {
+			return fmt.Errorf("could not determine epoch transition metrics/events for finalized block: %w", err)
+		}
+		metrics = append(metrics, epochTransitionMetrics...)
+		events = append(events, epochTransitionEvents...)
 	}
 
-	// if EECC is triggered, update metric
+	// if epoch emergency fallback is triggered, update metric
 	if epochFallbackTriggered {
-		m.metrics.EpochEmergencyFallbackTriggered()
+		metrics = append(metrics, m.metrics.EpochEmergencyFallbackTriggered)
```

Comment:
I feel this logic should be consolidated with this one here: https://github.com/onflow/flow-go/blob/ca33e9e09c6b2bf2fe8edb71a3a5504d850fbc0a/state/protocol/badger/mutator.go#L541-L544

in that way, we append events and metrics updates together, same as [for the code path with `!epochFallbackTriggered`](https://github.com/onflow/flow-go/blob/ca33e9e09c6b2bf2fe8edb71a3a5504d850fbc0a/state/protocol/badger/mutator.go#L551-L565)

Reply:
Good point. 

The reason I had this case separated was to ensure the metric still gets set properly following a restart (where we would not enter the first block on line 541). However, this is already handled in the initialization code:
https://github.com/onflow/flow-go/blob/c60050541bb5eaef19cad610232d562bd80e39de/state/protocol/badger/state.go#L627-L635

Updated in https://github.com/onflow/flow-go/pull/2625/commits/0b38f8e9c28670420942a94e41e0805457cf4252